### PR TITLE
feat(snyk): Snyk Source plugin

### PR DESCRIPTION
**Pull request by candiduslynx**

### CODE REVIEW

File: plugins/source/snyk/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Tailscale Plugin
```
Comment (disq) [inline plugins/source/snyk/README.md:1]:
```suggestion
# Synk Plugin
```

File: plugins/source/snyk/README.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+
+The CloudQuery Tailscale plugin pulls configuration out of Tailscale and loads it into any supported CloudQuery destination (e.g. PostgreSQL).
```
Comment (disq) [inline plugins/source/snyk/README.md:3]: Synk

File: plugins/source/snyk/README.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+
+## Links
+
+- [User Guide](https://docs.cloudquery.io/docs/plugins/sources/tailscale/overview)
```
Comment (disq) [inline plugins/source/snyk/README.md:7]:
```suggestion
- [User Guide](https://docs.cloudquery.io/docs/plugins/sources/synk/overview)
```

File: plugins/source/snyk/client/client.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	OrganizationID string
+	organizations  []string
+
+	logger zerolog.Logger
+}
+
+var _ schema.ClientMeta = (*Client)(nil)
+
+func (c *Client) ID() string {
+	return "snyk/" + c.OrganizationID
```
Comment (disq) [inline plugins/source/snyk/client/client.go:25]:
```suggestion
	return c.OrganizationID
```
Unless OrgID can be empty...

File: plugins/source/snyk/client/spec.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	// EndpointURL is optional parameter to override the API URL for snyk.Client.
+	EndpointURL string `json:"endpoint_url,omitempty"`
+}
+
+const (
+	EnvSnykApiKey = "SNYK_API_KEY"
+)
+
+func (s *Spec) getClient(version string) (*snyk.Client, error) {
+	if len(s.APIKey) == 0 {
```
Comment (disq) [inline plugins/source/snyk/client/spec.go:27]:
```suggestion
	if s.APIKey == "" {
```

Checking for empty string using len is a bit controversial

File: plugins/source/snyk/main.go [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package main
+
+import (
+	"github.com/cloudquery/cloudquery/plugins/source/snyk/resources/plugin"
+	"github.com/cloudquery/plugin-sdk/serve"
+)
+
+const sentryDSN = "TODO"
```
Comment (disq) [inline plugins/source/snyk/main.go:8]:
```suggestion
const sentryDSN = "https://c2101fe3c5fd4f91a095b2b37dc6364a@o1396617.ingest.sentry.io/4504333793165313"
```

File: website/pages/docs/plugins/sources/overview.mdx [hunk 7]
```diff
@@ -30,6 +30,7 @@ Official source plugins follow [release stages](#source-plugin-release-stages).
             { name: "Datadog", stage:"Preview" },
             { name: "Tailscale", stage:"Preview" },
             { name: "Vercel", stage:"Preview" },
+            { name: "Snyk", stage:"Preview" },
```
Comment (disq) [inline website/pages/docs/plugins/sources/overview.mdx:33]: Add it after Datadog so preview plugins are sorted as a subset?

**Review by disq — CHANGES_REQUESTED** [review]

File: plugins/source/snyk/README.md (hunk 1 above)
Comment (candiduslynx) [reply plugins/source/snyk/README.md:1]: Better yet, `Snyk` =)

File: plugins/source/snyk/client/spec.go (hunk 5 above)
Comment (candiduslynx) [reply plugins/source/snyk/client/spec.go:27]: well, len is built into the underlying struct, so it's even faster that comparing to empty string IMO

File: website/pages/docs/plugins/sources/overview.mdx (hunk 7 above)
Comment (candiduslynx) [reply website/pages/docs/plugins/sources/overview.mdx:33]: then whole preview part still will require sorting (cloudflare w=needs to go on top for that part?)

File: plugins/source/snyk/.pre-commit-config.yaml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+repos:
```
Comment (erezrokah) [inline plugins/source/snyk/.pre-commit-config.yaml:1]: We can delete this file see https://github.com/cloudquery/cloudquery/pull/5559

File: plugins/source/snyk/client/client.go (hunk 4 above)
Comment (disq) [reply plugins/source/snyk/client/client.go:25]: This is also tied to the spec.Name sometimes in some plugins.

Comment (candiduslynx) [reply plugins/source/snyk/client/client.go:25]: orgID shouldn't be empty (only initially?), so will change

File: plugins/source/snyk/main.go (hunk 6 above)
Comment (disq) [inline plugins/source/snyk/main.go:8]: This is left as `TODO`, did you force push without pulling?

Comment (candiduslynx) [reply plugins/source/snyk/main.go:8]: ooh

Comment (candiduslynx) [reply plugins/source/snyk/main.go]: fixed

File: .github/pr_labeler.yml [hunk 9]
```diff
@@ -1,3 +1,9 @@
+cli:
```
Comment (erezrokah) [inline .github/pr_labeler.yml:1]: I'm not sure this change is needed. If we do this it creates the impression that you need to add the entry sorted/at the right place. Also doesn't seem related to this PR

File: website/versions/source-snyk.json [hunk 10]
```diff
@@ -0,0 +1 @@
+{ "latest": "plugins-source-snyk-v1.0.0" }
```
Comment (erezrokah) [inline website/versions/source-snyk.json:1]: I don't think we need to add this file, it will be generated by our CI

**Review by erezrokah — APPROVED** [review]
Looks good. Commented on the CI stuff the @disq already reviewed the code.

I do think we should not sort the `pr_labeler.yml` entries, as it only creates more work for developers. Also I think we don't need the comments `# destination plugins` and `# source plugins` as the entries have `plugins/source/*` or `plugins/destination/*` in the path so it's pretty clear what's the purpose of them

File: .github/pr_labeler.yml (hunk 9 above)
Comment (candiduslynx) [reply .github/pr_labeler.yml:1]: will revert then

File: website/versions/source-snyk.json (hunk 10 above)
Comment (candiduslynx) [reply website/versions/source-snyk.json]: removed

**Review by disq — APPROVED** [review]

File: plugins/source/snyk/client/spec.go [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+		s.APIKey = os.Getenv(EnvSnykApiKey)
```
Comment (yevgenypats) [inline plugins/source/snyk/client/spec.go:28]: Can we not read this via the env variable but only support it via `spec` as the spec reader already supports environment variable expansion and file expansion. This way the documentation is consistent with all other plugins.

**Review by yevgenypats — CHANGES_REQUESTED** [review]
I don't think I found the `dependency` API. Can you point me to that? Without it the plugin doesn't have any use-case as far as Im aware.

File: plugins/source/snyk/client/spec.go (hunk 11 above)
Comment (candiduslynx) [reply plugins/source/snyk/client/spec.go:28]: Tailscale & GitHub plugins have this, too. Should we remove this from them as well?

Comment (candiduslynx) [reply plugins/source/snyk/client/spec.go]: > I don't think I found the `dependency` API. Can you point me to that? Without it the plugin doesn't have any use-case as far as Im aware.

Currently there's no Snyk SDK that supports it.
We can either fork the one I'm using and implement it ourselves, or postpone.

Comment (yevgenypats) [reply plugins/source/snyk/client/spec.go]: Yes, we should remove this from tailscale and github. then either fork or use directly but that's the only interesting API in snyk

**Review by yevgenypats — APPROVED** [review]
